x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetReader.java

```diff
@@ -217,4 +230,267 @@ Bits getWrappedLiveDocs() {
         return in.getLiveDocs();
     }
 
+    @Override
+    public Terms terms(String field) throws IOException {
+        Terms t = super.terms(field);
+        if (strictTermsEnum == false || null == t) {
+            return t;
+        } else{
+            return new DocumentSafeTerms(t);
+        }
+    }
+
+    @Override
+    public SortedDocValues getSortedDocValues(String field) throws IOException {
+        final SortedDocValues sortedDocValues = in.getSortedDocValues(field);
+        if (strictTermsEnum == false) {
+            return sortedDocValues;
+        } else {
+            final Bits liveDocs = getLiveDocs();
+            final TermsEnum invertedIndexTermsEnum = getInvertedIndexTermsEnum(field);
+            return new DocumentSafeSortedDocValues(sortedDocValues, invertedIndexTermsEnum, liveDocs);
+        }
+    }
+
+    @Override
+    public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+        final SortedSetDocValues sortedSetDocValues = in.getSortedSetDocValues(field);
+        if (strictTermsEnum == false) {
+            return sortedSetDocValues;
+        } else {
+            final Bits liveDocs = getLiveDocs();
+            final TermsEnum invertedIndexTermsEnum = getInvertedIndexTermsEnum(field);
+            return new DocumentSafeSortedSetDocValues(sortedSetDocValues, invertedIndexTermsEnum, liveDocs);
+        }
+    }
+
+    private TermsEnum getInvertedIndexTermsEnum(String field) throws IOException{
+        final Terms terms = in.terms(field);
+        final TermsEnum invertedIndexTermsEnum;
+        if (terms == null) {
+            invertedIndexTermsEnum = null;
+        } else {
+            invertedIndexTermsEnum = terms.iterator();
+        }
+        return invertedIndexTermsEnum;
+    }
+
+    static class DocumentSafeTerms extends FilterTerms {
+
+        DocumentSafeTerms(Terms in) {
+            super(in);
+        }
+
+        @Override
+        public TermsEnum iterator() throws IOException {
+            return new DocumentSafeTermsEnum(in.iterator());
+        }
+    }
+
+    static class DocumentSafeTermsEnum extends FilterTermsEnum {
+
+        DocumentSafeTermsEnum(TermsEnum in) {
+            super(in);
+        }
+
+        @Override
+        public SeekStatus seekCeil(BytesRef term) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public BytesRef next() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void seekExact(long ord) throws IOException {
+            throw new UnsupportedOperationException();
+        }
```

Comment:
It'b be nice to have a good error message in these UnsupportedOperationExceptions explaining why they are disallowed.